Add UnbufferedSerial and BufferedSerial API documentation

hugueskamba opened the pull request:
## Merge only after https://github.com/ARMmbed/mbed-os-examples-docs_only/pull/66
## The example code snippet URLs need to be amended prior to merging!

hugueskamba committed:
Add UnbufferedSerial and BufferedSerial API documentation

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+
+## Configuration
+
+The following parameters can be configured at object instantiation:
```
I would re-use most of the text from https://os.mbed.com/docs/mbed-os/v5.15/apis/rawserial.html to describe the different parameters. It was a nice explanation so worth keeping.

hugueskamba replied:
The explanation of the parameters is still there.

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+* Start bit: Indicates the start of UART data transmission
+* Data frame: Can be 5 to 8 (or 9 if a parity bit is not used) bits long for the actual data being transferred.
+* Parity bit: Optional bit, used for data error detection.
+* Stop bits: Can be 1 to 2 bits long to signal the end of a data packet.
```
The section from `TX and RX pins can be specified as` until `Stop bits` is quite redundant with the `Configuration` section later on so I would move and combine this section with `Configuration`.

hugueskamba replied:
I do not agree. The section you mentioned talks about UART in general.
The configuration section talks about what can be configured using the BufferedSerial class.
If someone already knows about UART he can have a quick look at the uncluttered configuration section to see what can be configured and what the default configuration is. The upper section is more for people who may not be familiar with UART.

evedon replied:
That's fine.

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+* Parity bit: Optional bit, used for data error detection.
+* Stop bits: Can be 1 to 2 bits long to signal the end of a data packet.
+
+The `BufferedSerial` class maps targets underlying serial communication functions. See the [porting guide](../porting/serial-port.html) for target serial support.
```
I am not sure if this sentence is useful. `BufferedSerial` and `UnbufferedSerial` provide an API that calls the underlying HAl API functions. Perhaps try to rephrase.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os-5-docs/compare/b14e6f622f7fbb29482b769636c4c4ab4b513ee5..934953559ca74a40ae8258e144792d1f2efc79af

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+
+When the RX interrupt is trigged, the `BufferedSerial` class stores the byte(s) available to read from the hardware buffer to an internal intermediary buffer. When a read request is made, the `BufferedSerial` class uses a mutex lock and enters a critical section to read out the number of bytes requested if as many are available in the intermediary buffer.
+To transmit multiple bytes, the class uses an intermediary buffer to store the bytes to send and monitors the serial interface to transfer them to the hardware buffer as soon as it is available. However, all bytes are written unbuffered if in a critical section.
+Using intermediary buffers allows it to be used reliably for input from non-interrupt context whilst avoiding excess spinning waiting for transmission buffer space.
```
Good section. Perhaps you could explain how the buffers are curently configured.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os-5-docs/compare/b14e6f622f7fbb29482b769636c4c4ab4b513ee5..934953559ca74a40ae8258e144792d1f2efc79af

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+
+## Example
+
+[![View code](https://www.mbed.com/embed/?url=https://os.mbed.com/teams/mbed_example/code/BufferedSerial/)](https://os.mbed.com/teams/mbed_example/code/BufferedSerial/file/112a40a5991a/main.cpp)
```
It would be good to provide multiple snippets like Serial and RawSerial did.
Perhaps you could comment specifically on printing using `printf` and show code snippet since there were many previsus examples on how to do it from a serial class.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os-5-docs/compare/b14e6f622f7fbb29482b769636c4c4ab4b513ee5..934953559ca74a40ae8258e144792d1f2efc79af

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,58 @@
+
+### Mbed OS usage
+
+Common use cases for `BufferedSerial` are IRQ heavy UART operations, such as the [BLE cordio in the transport driver](https://github.com/ARMmbed/mbed-os/blob/master/features/FEATURE_BLE/targets/TARGET_CORDIO/driver/H4TransportDriver.cpp#L62).
```
This is for Unbuffered I think.
Perhps say that this is used in retarget code for console printing.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os-5-docs/compare/b14e6f622f7fbb29482b769636c4c4ab4b513ee5..934953559ca74a40ae8258e144792d1f2efc79af

evedon commented on docs/api/io/UnbufferedSerial.md:
```diff
@@ -0,0 +1,24 @@
+
+The `UnbufferedSerial` class provides UART functionality with an API similar to the [`BufferedSerial`](./BufferedSerial.md) class. The classes also share the same default configurations.
+
+Unlike the `BufferedSerial` class, the `UnbufferedSerial` class does not use intermediary buffers to store bytes to transmit to or read from the hardware. The user application is responsible for processing each byte as they are received. The method to read data returns only one byte for every call. It is therefore suitable when more control is required and for use in interrupt handlers with the R ...
```
Perhaps also add that it has no locks so it is safe to use in interrupts provided that it is the only instance that is using that serial port.

hugueskamba replied:
https://github.com/ARMmbed/mbed-os-5-docs/compare/934953559ca74a40ae8258e144792d1f2efc79af..32c308336088d2e5634585b28f720375c550dad7

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+TX and RX pins can be specified as Not Connected (NC) for simplex (unidirectional) communication or as valid pins for full duplex (bi-directional) communication.
+
+Data is sent/received at a pre-defined speed called baud rate. Standard baud rate include 9600, 119200, 115200 or others.
```
```suggestion
`Baud rate` - Data is sent and received at a pre-defined speed called baud rate. Standard baud rates include 9600, 119200, 115200 or others.
```

hugueskamba replied:
That `Baud rate  -` does not flow with the rest. I am taking the other changes.

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+The `BufferedSerial` class provides UART functionality, it is the recommended class for serial data transfers. It allows sending and receiving bytes of data in a sequence using separate transmit (TX) and receive pins (RX). Communication can be between two processors or for sending text to a console.
+
+TX and RX pins can be specified as Not Connected (NC) for simplex (unidirectional) communication or as valid pins for full duplex (bi-directional) communication.
```
```suggestion
Serial channels have the following characteristics:

`TX and RX pins` can be specified as Not Connected (NC) for simplex (unidirectional) communication or as valid pins for full duplex (bi-directional) communication.
```

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+Data is sent/received at a pre-defined speed called baud rate. Standard baud rate include 9600, 119200, 115200 or others.
+
+Pieces of data are transmitted using data packets of configurable sizes divided in different sections which include:
```
```suggestion
Data are transmitted using packets of configurable sizes divided in different sections which include:
```

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+Pieces of data are transmitted using data packets of configurable sizes divided in different sections which include:
+
+* Start bit: Indicates the start of UART data transmission
```
```suggestion
* Start bit: Indicates the start of UART data transmission.
```

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+  - _TX_
+  - _RX_
+  - _Baud rate_ - defaults to the value configured in `mbed-os/platform/mbed_lib.json`.
```
It would be nicer to write the default as well (rather than asking the user to check it in a file). In any case you mentioni it line 42.

hugueskamba replied:
If we happen to change them we will have to modify this also.

hugueskamba replied:
There is no default for TX or RX. There is a default for the rest and I mention it below. I do not see what to add here.

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+  - _Baud rate_
+  - _Data frame length_
+  - _Parity bit_
+  - _Stop bits_
```
It would be clearer to write the default as well. You do menation it on the next line as `9600-8-N-1` but it would make the notation more understandable.

hugueskamba replied:
I prefer the bullet point to be uncluttered. I have added a bit of text to explain the common notation.

evedon commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,71 @@
+
+Mbed OS redefines target-dependent I/O functions in the C library to allow the C standard I/O library functions (`s\v\f\n\printf`, `scanf`, etc) to be used in user application for printing to the console.
+
+The [system I/O retarget code](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/mbed_retarget.cpp) can be configured to be buffered or unbuffered depending on the configuration in [`platform/mbed_lib.json`](https://github.com/ARMmbed/mbed-os/blob/master/platform/mbed_lib.json). When it is buffered, [an instance of a `BufferedSerial` class](https://github.com/ARMmbed/mbed-os/blob/mas ...
```
I am not keen on having so many links. It might become a maintenance nightmare. 
Also, I would list the configuration parameters from `platform/mbed_lib.json` that are applicable to thsi section.

evedon commented on docs/api/io/UnbufferedSerial.md:
```diff
@@ -0,0 +1,24 @@
+<span class="images">![](https://os.mbed.com/docs/development/mbed-os-api-doxy/classmbed_1_1_unbuffered_serial.png)<span>`UnbufferedSerial` class hierarchy</span></span>
+
+The `UnbufferedSerial` class provides UART functionality with an API similar to the [`BufferedSerial`](./BufferedSerial.md) class. The classes also share the same default configurations.
+
```
You could re-emphasize here that for normal blocking application that require a serial channel for something other than the console, BufferedSerial will perform better than UnbufferedSerial and will cause less CPU load and latency issues.
UnbufferedSerial should only be used by applications that are short of RAM and can't afford buffering or need more control of the serial port and use it from IRQ.

Commit:
Edit BufferedSerial.md

Make initial copy edits, mostly for active voice, correct links and consistent style.

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+
+<span class="images">![](https://os.mbed.com/docs/development/mbed-os-api-doxy/classmbed_1_1_buffered_serial.png)<span>`BufferedSerial` class hierarchy</span></span>
+
+The `BufferedSerial` class provides UART functionality. We recommend you use this class for serial data transfers. You can use it to send and receive bytes of data in a sequence using separate transmit (TX) and receive pins (RX). Communication can be between two processors or for sending text to a console.
```
What does this last sentence mean?

hugueskamba replied:
It means that a device can interface with another device (such as sensors, printers, or another board) to exchange data or to send text to be displayed on a text-based computer interface.

evedon replied:
I would either remove the original sentence or replace it with the explanation in your comment.

hugueskamba replied:
Thanks. [Done](https://github.com/ARMmbed/mbed-os-5-docs/pull/1220/commits/0c90ed33cc75aadc0b1a82485ffd0ba90316daa1).

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+
+Serial channels have the following characteristics:
+
+- TX and RX pins - can be specified as Not Connected (NC) for simplex (unidirectional) communication or as valid pins for full duplex (bidirectional) communication.
```
Can be specified by whom or by what? You, the user?

evedon replied:
I suggest the text from RawSerial: `The physical serial transmit and receive pins. You can specify a TX or RX pin as Not Connected (NC) to get simplex communication, or specify both to get full duplex communication.`

hugueskamba replied:
> Can be specified by whom or by what? You, the user?

The user can.

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+Serial channels have the following characteristics:
+
+- TX and RX pins - can be specified as Not Connected (NC) for simplex (unidirectional) communication or as valid pins for full duplex (bidirectional) communication.
+- Baud rate - predefined speed at which data is sent and received. Standard baud rates include 9600, 119200 and 115200.
```
Who or what sends the data? The BufferedSerial class?

hugueskamba replied:
UART

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+
+The `BufferedSerial` calls the underlying HAL API functions. Please see the [porting guide](../porting/serial-port.html) for target serial support.
+
+When the RX interrupt is trigged, the `BufferedSerial` class stores the byte(s) available to read from the hardware buffer to an internal intermediary buffer. When a read request is made, the `BufferedSerial` class uses a mutex lock and enters a critical section to read out the number of bytes requested if as many are available in the intermediary buffer.
```
What do we mean by "the RX is trigged," and who or what trigged it?
Who or what makes the read request?

hugueskamba replied:
`RX` is explained in the first paragraph of the page. It is triggered when receiving data from whatever is connected to that interface (a sensor, a console, etc...).

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+When the RX interrupt is trigged, the `BufferedSerial` class stores the byte(s) available to read from the hardware buffer to an internal intermediary buffer. When a read request is made, the `BufferedSerial` class uses a mutex lock and enters a critical section to read out the number of bytes requested if as many are available in the intermediary buffer.
+
+To transmit multiple bytes, the class uses an intermediary buffer to store the bytes to send and monitors the serial interface to transfer them to the hardware buffer as soon as it is available. However, all bytes are written unbuffered if in a critical section.
+Using intermediary buffers allows it to be used reliably for input from noninterrupt context while avoiding excess spinning waiting for transmission buffer space.
```
Using intermediary buffers allows us to use what reliably? The class?

hugueskamba replied:
The UART interface.

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+
+The RX and TX buffers are circular buffers with preallocated sizes configurable using the configuration parameters `uart-serial-rxbuf-size` and `uart-serial-txbuf-size`. You can find both configuration parameters in `drivers/mbed_lib.json`.
+
+## Configuration
```
Are these the same configuration parameters you get by running the `--config -v` command?

hugueskamba replied:
This is to configure a UART interface via the BufferedSerial class for the user application.
Nothing to do with Mbed CLI.

AnotherButler commented on docs/api/io/BufferedSerial.md:
```diff
@@ -0,0 +1,73 @@
+
+Mbed OS redefines target-dependent I/O functions in the C library to allow you to use the C standard I/O library functions (`s\v\f\n\printf`, `scanf` and so on) in your application for printing to the console.
+
+You can refigure the system I/O retarget code to be buffered or unbuffered, depending on the configuration of the parameter `stdio-buffered-serial` in `platform/mbed_lib.json`. When it is buffered, the retarget code uses an instance of a `BufferedSerial` class to perform the actual printing. This is where `BufferedSerial`'s `Filehandle` inheritance is used.
```
Who or what uses the `Filehandle` inheritance?

hugueskamba replied:
It says `BufferedSerial's`.
You see in the diagram at the top of the page that `BufferedSerial` inherits from `FileHandle`. Because it inherits from `FileHandle` it can be used for the re-target code.

AnotherButler reviewed (commented):
I've left some edits and queries. Please review my changes to make sure I didn't accidentally change the meaning of anything.

Commit:
Edit UnbufferedSerial.md

Edit file, mostly for accurate links and active voice.

hugueskamba commented on docs/api/io/BufferedSerial.md:
```diff
@@ -2,74 +2,72 @@
 You can view more information about the configurable settings and functions in the class reference.
 
-## Class reference
+## BufferedSerial class reference
```
This edit seems redundant since the page is about `BufferedSerial`.

hugueskamba commented on docs/api/io/BufferedSerial.md:
```diff
@@ -2,74 +2,72 @@
-[![View code](https://www.mbed.com/embed/?url=https://os.mbed.com/teams/mbed_example/code/BufferedSerial/)](https://os.mbed.com/teams/mbed_example/code/BufferedSerial/file/112a40a5991a/main.cpp)
+<span class="notes">**Note**: On a Windows machine, you need to install a USB serial driver. Please see the [Windows serial configuration](../tutorials/serial-communication.html#windows-serial-driver) document for more information.</span>
 
+## BufferedSerial examples
```
This edit seems redundant since the page is about `BufferedSerial`.

hugueskamba committed:
Edits based on feedback.

hugueskamba commented:
> I've left some edits and queries. Please review my changes to make sure I didn't accidentally change the meaning of anything.

I reviewed the edits, and commented on some of them.
I have replied to the review comments and pushed some changes to address them.

hugueskamba committed:
Clarify possible communications with UART interface

evedon commented:
@AnotherButler Please review after the last edits. Note that the code for the examples has been merged so this PR is ready for merge once you are happy with it.

Commit:
Fix typos in BufferedSerial.md

Delete extra space and extra comma.
